### PR TITLE
Do not catch and ignore fatal exceptions

### PR DESCRIPTION
**Pull request by pjfanning**
I was pitching for this in https://github.com/FasterXML/jackson-databind/pull/3826

I don't think it is a good idea to risk messing with fatal errors like OutOfMemoryErrors.

If this util seems ok, it could in theory be moved to jackson-core and used in other Jackson modules.

fyi @JooHyukKim

### CODE REVIEW

File: src/main/java/com/fasterxml/jackson/databind/util/ExceptionUtil.java [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+     * <a href="https://www.scala-lang.org/api/2.13.10/scala/util/control/NonFatal$.html">scala.util.control.NonFatal</a>.
+     *
+     * @param throwable to check
+     * @return whether the <code>Throwable</code> is a fatal error
+     */
+    public static boolean isFatal(Throwable throwable) {
+        return (throwable instanceof VirtualMachineError
+                || throwable instanceof ThreadDeath
+                || throwable instanceof InterruptedException
+                || throwable instanceof LinkageError);
```
Comment (cowtowncoder) [inline src/main/java/com/fasterxml/jackson/databind/util/ExceptionUtil.java:23]: I am not sure `LinkageError` should be included in this list: isn't it one of failures that specifically can occur on various platforms trying to load optional extensions?

**Comment by cowtowncoder** [comment]
Ok. I guess that's fine, except for my concern wrt `LinkageError`.

But how about adding convenience method of `rethrowIfFatal(t)` to reduce boilerplate code on caller?

File: src/main/java/com/fasterxml/jackson/databind/util/ExceptionUtil.java (hunk 1 above)
Comment (pjfanning) [reply src/main/java/com/fasterxml/jackson/databind/util/ExceptionUtil.java:23]: I wouldn't be against removing LinkageError here or being specific. Some LinkageErrors seem more fatal than others - see the subclasses here - https://docs.oracle.com/javase/7/docs/api/java/lang/LinkageError.html

NoClassDefError seems least fatal but in use cases where I added this util to be used, we tend to have the Class already.

**Comment by pjfanning** [comment]
> Ok. I guess that's fine, except for my concern wrt `LinkageError`.
> 
> But how about adding convenience method of `rethrowIfFatal(t)` to reduce boilerplate code on caller?

I've refactored to use `rethrowIfFatal(t)`.

I still haven't changed LinkageError.

File: src/main/java/com/fasterxml/jackson/databind/util/ExceptionUtil.java (hunk 1 above)
Comment (cowtowncoder) [reply src/main/java/com/fasterxml/jackson/databind/util/ExceptionUtil.java]: Yeah the concern I have is specifically that the optional loading should not die on missing classes; that's pretty much the main use case. I wish actual test case. But conversely I haven't seen reports where catching one of fatal errors here have been problematic -- theoretically I could see `InterruptedException` catching being problematic, of these I think.

**Comment by pjfanning** [comment]
I've changed the 'fatal' check to ignore NoClassDefError

File: src/main/java/com/fasterxml/jackson/databind/util/ExceptionUtil.java [hunk 2]
```diff
@@ -49,6 +49,12 @@ private static boolean isFatal(Throwable throwable) {
         return (throwable instanceof VirtualMachineError
                 || throwable instanceof ThreadDeath
                 || throwable instanceof InterruptedException
-                || throwable instanceof LinkageError);
+                || throwable instanceof ClassCircularityError
+                || throwable instanceof ClassFormatError
+                || throwable instanceof ExceptionInInitializerError
```
Comment (cowtowncoder) [inline src/main/java/com/fasterxml/jackson/databind/util/ExceptionUtil.java:54]: Also not sure this belongs here -- isn't this a wrapper used around other failures, during class initialization?
And as such likely failure mode when trying to load something that depends on something else not available.

Others make sense.

Comment (pjfanning) [reply src/main/java/com/fasterxml/jackson/databind/util/ExceptionUtil.java:54]: which Error is not fatal? all the new ones or one in particular?

**Review by cowtowncoder — APPROVED** [review]

**Comment by cowtowncoder** [comment]
Ok looks good. Will merge.

**Pull request merged**